File: desktop/src/main/java/com/gengzi/desktop/ocr/OcrService.java
package com.gengzi.desktop.ocr;

import com.gengzi.desktop.llm.BackendClient;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.function.Consumer;

/**
 * OCR服务（实际使用LLM进行图片分析）
 * 支持流式和非流式两种模式
 */
public class OcrService {
    private final BackendClient backendClient;

    public OcrService(String tesseractPath, String tessdataPath, String language, BackendClient backendClient) {
        this.backendClient = backendClient;
    }

    public OcrService(BackendClient backendClient) {
        this.backendClient = backendClient;
    }

    /**
     * 分析图片并返回完整答案（非流式）
     * @param image 截图
     * @return LLM分析结果
     */
    public String recognize(BufferedImage image) throws Exception {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ImageIO.write(image, "png", baos);
        byte[] imageData = baos.toByteArray();

        return backendClient.solveWithImage(imageData, "Please analyze this image and answer any questions shown in it. Provide a concise, correct answer.");
    }

    /**
     * 分析图片并流式返回答案（流式）
     * @param image 截图
     * @param chunkConsumer 接收每个文本块的回调函数
     */
    public void recognizeStream(BufferedImage image, Consumer<String> chunkConsumer) throws Exception {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ImageIO.write(image, "png", baos);
        byte[] imageData = baos.toByteArray();

        backendClient.solveWithImageStream(imageData, "Please analyze this image and answer any questions shown in it. Provide a concise, correct answer.", chunkConsumer);
    }
}

File: python_ui/ocr_service.py
from io import BytesIO
from PIL import Image


MAX_DIMENSION = 1280
JPEG_QUALITY = 70


class OcrService:
    def __init__(self, backend_client):
        self._backend = backend_client

    def recognize(self, image):
        payload = self._encode_image(image)
        return self._backend.solve_with_image(
            payload,
            "Please analyze this image and answer any questions shown in it. Provide a concise, correct answer.",
        )

    def recognize_stream(self, image, chunk_consumer):
        payload = self._encode_image(image)
        return self._backend.solve_with_image_stream(
            payload,
            "Please analyze this image and answer any questions shown in it. Provide a concise, correct answer.",
            chunk_consumer,
        )

    def _encode_image(self, image):
        normalized = self._normalize_image(image)
        scaled = self._scale_down(normalized)
        out = BytesIO()
        scaled.save(out, format="JPEG", quality=JPEG_QUALITY, optimize=True)
        return out.getvalue()

    def _normalize_image(self, image):
        if image.mode == "RGB":
            return image
        return image.convert("RGB")

    def _scale_down(self, image):
        width, height = image.size
        max_dim = max(width, height)
        if max_dim <= MAX_DIMENSION:
            return image
        scale = MAX_DIMENSION / float(max_dim)
        new_size = (max(1, int(round(width * scale))), max(1, int(round(height * scale))))
        return image.resize(new_size, Image.BILINEAR)
